src/sema/function_annotation.rs

```diff
@@ -1,18 +1,131 @@
+        match annotation.id.name.as_str() {
+            "selector" => function_selector(func, annotation, &mut diagnostics, ns),
+            _ if !func.has_body => {
+                diagnostics.push(Diagnostic::error(
+                    annotation.loc,
+                    format!(
+                        "annotation {} not allowed on {} with no body",
+                        annotation.id.name, func.ty
+                    ),
+                ));
+            }
+            // handled in function_body_annotations
+            _ => (),
+        }
+    }
+
+    ns.diagnostics.extend(diagnostics);
+}
+
+/// Parse the selector from an annotation and assign it to the function
+fn function_selector(
+    func: &mut Function,
+    annotation: &pt::Annotation,
+    diagnostics: &mut Diagnostics,
+    ns: &mut Namespace,
+) {
+    if func.ty != pt::FunctionTy::Function
+        && (!ns.target.is_substrate() || func.ty != pt::FunctionTy::Constructor)
+    {
+        diagnostics.push(Diagnostic::error(
+            annotation.loc,
+            format!("overriding selector not permitted on {}", func.ty),
+        ));
+        return;
+    }
+
+    if !func.is_public() {
+        diagnostics.push(Diagnostic::error(
+            annotation.loc,
+            format!(
+                "overriding selector only permitted on 'public' or 'external' function, not '{}'",
+                func.visibility
+            ),
+        ));
+        return;
+    }
+
+    if let Some((prev, _)) = &func.selector {
+        diagnostics.push(Diagnostic::error_with_note(
+            annotation.loc,
+            format!("duplicate @selector annotation for {}", func.ty),
+            *prev,
+            "previous @selector".into(),
+        ));
+        return;
+    }
+
+    match &annotation.value {
+        pt::Expression::ArrayLiteral(_, values) => {
+            let mut selector = Vec::new();
+
+            for expr in values {
+                let expr = match expr {
+                    pt::Expression::HexNumberLiteral(loc, n) => {
+                        hex_number_literal(loc, n, ns, diagnostics, ResolveTo::Type(&Type::Uint(8)))
+                    }
+                    pt::Expression::NumberLiteral(loc, base, exp) => number_literal(
+                        loc,
+                        base,
+                        exp,
+                        ns,
+                        &BigInt::one(),
+                        diagnostics,
+                        ResolveTo::Type(&Type::Uint(8)),
+                    ),
+                    _ => {
+                        diagnostics.push(Diagnostic::error(
+                            expr.loc(),
+                            "literal number expected".into(),
+                        ));
+                        continue;
+                    }
+                };
+
+                if let Ok(Expression::NumberLiteral(_, _, v)) = expr {
+                    selector.push(v.to_u8().unwrap());
+                }
```

Comment:
What happens when we don't enter this condition? Adding a `unreachable!()` to an `else` branch would be worth it.

Reply:
You inserted the `unreachable` in the wrong if-condition. I mentioned this one here, not the one below.

Reply:
Sorry, my code was really bad yesterday. :disappointed: